src/utils/usePromiseResolver.ts

```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import {isPromise} from '../utils/guards';
+
+export const usePromiseResolver = <T>(mayBePromise: T | PromiseLike<T>): T | null => {
+  const [resolved, setResolved] = React.useState(() => {
+    return isPromise(mayBePromise) ? null : mayBePromise
+  })
+
+  React.useEffect(() => {
+    if (!isPromise(mayBePromise)) return setResolved(mayBePromise)
+
+    let isMounted = true
+
+    setResolved(null)
+    mayBePromise.then(resolvedValue => {
+      if (isMounted) setResolved(resolvedValue)
+    }, () => undefined)
```

Comment:
question: why add `() => undefined` as the rejection handler? Don't we want to know if a promise is rejected?

Reply:
It depends on the hook purpose. If it's generic purpose hook, it should provide, otherwise in case of what provider does, it doesn't make sense as provider doesn't do anything with a rejection.